cpp/include/cuvs/neighbors/ann_mg.hpp

```diff
@@ -0,0 +1,1417 @@
+
+namespace cuvs::neighbors::ivf_pq {
+/// \ingroup ann_mg_cpp_index_params
+/**
+ * IVF-PQ MG build parameters struct
+ *
+ * Usage example:
+ * @code{.cpp}
+ * cuvs::neighbors::ivf_pq::mg_index_params index_params;
+ * index_params.mode = cuvs::neighbors::mg::distribution_mode::SHARDED;
+ * index_params.metric = cuvs::distance::DistanceType::L2Expanded;
+ * index_params.metric_arg = 2.0f;
+ * index_params.add_data_on_build = true;
+ * @endcode
+ *
+ */
+struct mg_index_params : public cuvs::neighbors::ivf_pq::index_params {
+  /** Distribution mode */
+  cuvs::neighbors::mg::distribution_mode mode;
+};
+}  // namespace cuvs::neighbors::ivf_pq
+
+namespace cuvs::neighbors::cagra {
+/// \ingroup ann_mg_cpp_index_params
+/**
+ * CAGRA MG build parameters struct
+ *
+ * Usage example:
+ * @code{.cpp}
+ * cuvs::neighbors::cagra::mg_index_params index_params;
+ * index_params.mode = cuvs::neighbors::mg::distribution_mode::SHARDED;
+ * index_params.metric = cuvs::distance::DistanceType::L2Expanded;
+ * index_params.metric_arg = 2.0f;
+ * index_params.add_data_on_build = true;
+ * @endcode
+ *
+ */
+struct mg_index_params : public cuvs::neighbors::cagra::index_params {
+  /** Distribution mode */
+  cuvs::neighbors::mg::distribution_mode mode;
+};
+}  // namespace cuvs::neighbors::cagra
+
+namespace cuvs::neighbors::mg {
+using pool_mr = rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>;
+
+/// \defgroup ann_mg_cpp_nccl_clique NCCL clique utility
+
+/// \ingroup ann_mg_cpp_nccl_clique
+struct nccl_clique {
+  /**
+   * Instantiates a NCCL clique with all available GPUs
+   *
+   * @param[in] percent_of_free_memory percentage of device memory to pre-allocate as memory pool
+   *
+   */
+  nccl_clique(int percent_of_free_memory = 80);
+
+  /**
+   * Instantiates a NCCL clique
+   *
+   * Usage example:
+   * @code{.cpp}
+   * int n_devices;
+   * cudaGetDeviceCount(&n_devices);
+   * std::vector<int> device_ids(n_devices);
+   * std::iota(device_ids.begin(), device_ids.end(), 0);
+   * cuvs::neighbors::mg::nccl_clique& clique(device_ids); // first device is the root rank
+   * @endcode
+   *
+   * @param[in] device_ids list of device IDs to be used to initiate the clique
+   * @param[in] percent_of_free_memory percentage of device memory to pre-allocate as memory pool
+   *
+   */
+  nccl_clique(const std::vector<int>& device_ids, int percent_of_free_memory = 80);
+
+  void nccl_clique_init();
+  const raft::device_resources& set_current_device_to_root_rank() const;
+  ~nccl_clique();
+
+  int root_rank_;
+  int num_ranks_;
+  int percent_of_free_memory_;
+  std::vector<int> device_ids_;
+  std::vector<ncclComm_t> nccl_comms_;
+  std::vector<std::unique_ptr<pool_mr>> per_device_pools_;
+  std::vector<raft::device_resources> device_resources_;
+};
+
+using namespace raft;
+
+template <typename AnnIndexType, typename T, typename IdxT>
+class ann_interface {
+ public:
+  template <typename Accessor>
+  void build(raft::resources const& handle,
```

Comment:
I'm not in love with having the build, search, extend functions as methods on the index. We have intentionally kept these as free functions so that the APIs are easier to handle and wrap in other languages. Please pull these back into free functions that accept the index object and have the index object only maintain simple needed state about the index (opaque to the user). 

I think we should be able to keep the free functions but still make the underlying APIs smart. We will need to build similar APIs for the tiered index also.